src/rule_types.jl

```diff
@@ -51,12 +51,15 @@ See also: [`frule`](@ref), [`rrule`](@ref), [`Rule`](@ref), [`DNERule`](@ref), [
 abstract type AbstractRule end
 
 # this ensures that consumers don't have to special-case rule destructuring
-Base.iterate(rule::AbstractRule) = (rule, nothing)
+Base.iterate(rule::AbstractRule) = (@warn "iterating rules is going away"; (rule, nothing))
 Base.iterate(::AbstractRule, ::Any) = nothing
 
 # This ensures we don't need to check whether the result of `rrule`/`frule` is a tuple
 # in order to get the `i`th rule (assuming it's 1)
-Base.getindex(rule::AbstractRule, i::Integer) = i == 1 ? rule : throw(BoundsError())
+function Base.getindex(rule::AbstractRule, i::Integer)
+    @warn "iterating rules is going away"
```

Comment:
We still need to figure out hwo to deprecate `AbstractRule`s altogether? I guess you'll figure that out when/once `acummulate!` etc get rewritten?

Reply:
I'm not too worried about deprecating them
as the changes are completely breaking.
And it isn't used in too many places.

But do need to workout how to remove them outright